### ISSUE TITLE
Newbie Question about Settings provider

### ISSUE THREAD

**Issue by clodo** [driftyco/ionic-starter-super#58]
Hi, what is the proper way to access the settings from other component?

I've put this in my component provider:

```javascript
import { Settings } from './settings';

@Injectable()
export class Affiliates {
  constructor(public http: Http, private settings: Settings) { }

  getApiUrlByTag(tagId) {
    return this.settings.load()
      .then(() => {
        return this.settings.getValue('option2') + tagId;
      });
  }
}
```

Also I had to change the `getValue` method in Settings provider to this:
```javascript
getValue(key: string) {
  // return this.storage.get(key);
  return this.settings[key];
}
```

The app works and I can get the setting parameters correctly, but I'm not sure it's the best way.

Thanks and sorry for the newbie question

**Comment by wbhob** [comment]
`this.settings.load()` does not take a promise, so that should not work. Just using the getter works find for me.

**Comment by clodo** [comment]
Hi @wbhob

My code works and if you look at the settings page the `ionViewWillEnter()` method has the same piece of code for build the form:
```javascript
this.settings.load().then(() => {
  this.settingsReady = true;
  this.options = this.settings.allSettings;

  this._buildForm();
});
```

**Comment by clodo** [comment]
Think I've got it, the problem is in the `getValue(key: string)` method.

```javascript
getValue(key: string) {
  return this.storage.get(key);
}
```

The `key` in the storage is `private SETTINGS_KEY: string = '_settings';`

```javascript
setValue(key: string, value: any) {
  this.settings[key] = value;
  return this.storage.set(this.SETTINGS_KEY, this.settings);
}
```

I then modified the `getValue(key: string)` method as follows to make it work:

```javascript
getValue(key: string) {
  return this.storage.get(this.SETTINGS_KEY)
    .then(settings => {
      return settings[key];
    });
}
```

**Comment by wbhob** [comment]
ah, good catch! I'll put that in right now. I will also add a callback method for people needing that.

**Comment by wbhob** [comment]
UPDATE: I added it to my PR #59, should be merged in the coming days.

**Comment by clodo** [comment]
Genial!
Another thing we needed is the initialization of settings because if you don't enter the settings page or call the `load()` method directly you cannot access the storage.

**Comment by wbhob** [comment]
Consider it done.

**Comment by clodo** [comment]
And you have a typo in your PR, the key for the storage index is `this.SETTINGS_KEY`

```javascript
getValue(key: string, callback?: (res: string) => void) {
  return this.storage.get(this.SETTINGS_KEY).then(settings => {
    if (callback) {
      callback(settings[key]);
    }
    else {
      return settings[key];
    }
  });
}
```

**Comment by wbhob** [comment]
That's actually how I found it... comitting now.

**Comment by wbhob** [comment]
Close this?

**Comment by bgies** [comment]
It certainly can't be a "Best Practice" to retrieve single values from the database every time you use them. It's certainly not in any other language I know of. Pulling from local storage is considered slow, so best practice would have to be pull the entire settings object once (probably in the constructor), and then pass them from there:

So, this code would be better 
`getValue(key: string) {
  // return this.storage.get(key);
  return this.settings[key];
}`

BUT... how to know when it's loaded?

**Comment by bgies** [comment]
Also, it seems that starting the setting key with an underscore prevents you from getting the values later... not sure why... but an underscore normally denotes private, so it doesn't seem to be a good choice for a database key. 

so ` private SETTINGS_KEY: string = '_settings';`  should be changed to this ` private SETTINGS_KEY: string = 'settings';` or  `private SETTINGS_KEY: string = 'myapp_settings';`

**Comment by bgies** [comment]
Ok... even I didn't like the tone of my previous comments after reading them over, so I thought I'd offer some constructive code. It might take me a few revisions to make it readable, so bear with me. Here's what I think is a much better pattern for the SettingsService: 

This file is just a namespace that exports all the variables that will NEVER change.
scr/app/constants/constants.ts
```
export namespace Constants { 
   //export class Constants {
   /* SETTINGS KEYS */
   export const STORAGE_KEY_SETTINGS: string = 'settings';
   export const STORAGE_KEY_USER: string = 'user';
   export const STORAGE_KEY_TOKEN: string = 'token';
   export const STORAGE_KEY_LOOKUPS: string = 'lookups';
   export const STORAGE_KEY_COMPANY: string = 'company';

   export const SERVER_URL: string = "http://localhost:8100";
   export const OPEN_SERVER_URL: string = "http://myapp.local:8000";
   export const DATA_SERVER_URL: string = "http://localhost:8100/api";  // NEEDS AUTHENTICATION

   export const LOOKUPS_STUB: string = "/lookup";
   export const COMPANY_URL_STUB: string = "/company";
   export const LOCATIONS_STUB: string = "/company/locationlist";
   export const USER_LOGIN_STUB: string = "/oauth/token";
}

```
Then the SettingsService.
src/providers/settings-service.ts

```
import { Injectable } from '@angular/core';
import { Storage } from '@ionic/storage';
import { Observable } from "rxjs"; 
import { Constants } from '../app/constants/constants';

/**
 * A simple settings/config class for storing key/value pairs with persistence.
 */
@Injectable()
export class SettingsService {
  settings: any = null;  // SET TO NULL TO MAKE CHECKING IT EASY

  settingsObservable: Observable<any> = 
     Observable.fromPromise(this.storage.get(Constants.STORAGE_KEY_SETTINGS)
           .then((value) => {
              if(value) {
                 this.settings = value;
                 return this._mergeDefaults(this._defaults);
              } else {
                 return this.setAll(this._defaults).then((val) => {
                    this.settings = val;
                 })
              }
           })
     ).share();

  _defaults: any;

  constructor(public storage: Storage, defaults: any) {
    this._defaults = defaults;
  }

  .... your other functions stay the same.... but load is deleted. 

  getValue(key: string) {
     return this.settings[key];
  }





```

Note that the Promise from Storage is converted to an observable AND SHARED. Shared Observables need a subscriber or they stay dormant, but because this is a Settings Services, we know we want it right away, so we'll call it in the app.component.ts file.... and then also attach to it from the First page (root page). 

In src/app/app.components.ts


```
..... other imports...... 
import { HttpModule } from '@angular/http';
import { SettingsService } from '../providers/settings-service';
import { TranslateService } from 'ng2-translate/ng2-translate';

..... other code...... 

constructor(
        platform: Platform,
        public translate: TranslateService,
        public storage: Storage,
        public httpModule: HttpModule
        public settings: SettingsService) {

      platform.ready().then(() => {
      // Okay, so the platform is ready and our plugins are available.
      // Here you can do any higher level native things you might need.
      StatusBar.styleDefault();
      Splashscreen.hide();
    });    
  }
 
  ngOnInit() {
     //this.storage.clear();   // FOR STARTING OVER AS I DEVELOP
     let that = this;    // NEVER TRUST this will be what you expect in inner functions, so keep a copy of it. 

    // I have code that gets the Browser language here, but omitted it for brevity
     
     this.settings.settingsObservable.subscribe((value) => {
        that.settings.settings['USER_LOGGED_IN'] = false;
        if (that.settings.settings['USER_LANGUAGE'] !== '') {
           that.translate.setDefaultLang(this.settings.settings['USER_LANGUAGE']);
           that.translate.use(this.settings.settings['USER_LANGUAGE']);
           // START LOADING OTHER SERVICES THAT DEPEND ON THE SETTINGSSERVICE
           that.lookupsService.load();
           that.companyService.load();

        }
     });
  }
}

```

OK, then in the First page that will load..... check to see if the SettingsService has finished loading, if not attach to the Observable, otherwise just go ahead and use the settings. ALL other pages can just use this.settings.getValue('the value I want'). 

src/pages/home/home.ts (or whatever your first page actually is)

```
import { SettingsService } from '../../providers/settings-service';

  stuffIWantToDoWhenSettingsAreAvailable() {
            if (that.settings.settings['USER_LANGUAGE'] == '') {
               console.log('HomePage ionViewDidLoad presenting LanguageModal');
               that.presentLanguageModal();
               return;
            }
            
            this.checkLoginNeeded();
       });

  }

  ionViewDidLoad() {
  
    let that = this;
    // if the settings haven't loaded yet, wait for them. 
    if (this.settings.settings == null) {
       this.settings.settingsObservable.subscribe((value) => {
           stuffIWantToDoWhenSettingsAreAvailable()  
          return;
    } 
   
   // Setting were already available so... just call it here.
   stuffIWantToDoWhenSettingsAreAvailable();
}

```

Basically the short version of what's happening is that the Storage call is converted from a promise to an observable, and shared... then called from the app.module.ts... in the first page that loads you check if settings is null, and use the observable if they aren't and call the setup function from there... if the settings are already loaded, just call the setup function. 

NOTE also that I didn't use _ as a prefix for my storage keys... 

I'm sure this post will be a formatting mess, so if anyone wants to make it look better, correct typos, or make it better, please feel free.

**Comment by wbhob** [comment]
Dude, thanks a ton! I'll add that to my PR later

**Comment by b0jan** [comment]
Much better implementation by @bgies. Something I think would be also very useful is broadcasting settings changes using [BehaviorSubject](https://github.com/Reactive-Extensions/RxJS/blob/master/doc/api/subjects/behaviorsubject.md) or maybe ReplaySubject, like this (not sure if this is the best implementation but it works):

```
import { Injectable } from '@angular/core';
import { Constants } from '../constants/constants';
import { Observable, BehaviorSubject }  from 'rxjs';
import { Storage } from '@ionic/storage';

@Injectable()
export class Settings {

  private settings: any = null;  // SET TO NULL TO MAKE CHECKING IT EASY
  private settingsSub: BehaviorSubject<any> = new BehaviorSubject(this.settings);
  private settingsKey: string = Constants.STORAGE_KEY_SETTINGS;

  settingsObservable: Observable<any> =
     Observable.fromPromise(this.storage.get(this.settingsKey)
           .then(value => {
              if(value) {
                this.settingsSub.next(value);
                 this.settings = value;
                 return this._mergeDefaults(this._defaults);

              } else {
                this.settingsSub.next(this._defaults);
                return this.setAll(this._defaults).then(val => {
                  this.settings = val;
                })
              }
           })
     ).share();

  _defaults: any;

  constructor(private storage: Storage, private defaults: any) {
    this._defaults = defaults;
  }

  _mergeDefaults(defaults: any) {
    for(let k in defaults) {
      if(!(k in this.settings)) {
        this.settings[k] = defaults[k];
      }
    }
    return this.setAll(this.settings);
  }

  merge(settings: any) {
    for(let k in settings) {
      this.settings[k] = settings[k];
    }
    return this.save();
  }

  setValue(key: string, value: string | number) {
    this.settings[key] = value; // Update value in settings object
    this.settingsSub.next(this.settings); // Alert subscribers
    return this.storage.set(this.settingsKey, this.settings); // Save to localStorage
  }

  setAll(value: any) {
    return this.storage.set(this.settingsKey, value);
  }

  getValue(key: string) {
     return this.settings[key];
  }

  save() {
    return this.setAll(this.settings);
  }

  get allSettings() {
    return this.settings;
  }

  radio(): Observable<any> {
    return this.settingsSub.asObservable();
  }
}
```

**Comment by bgies** [comment]
I agree with @b0jan. Broadcasting settings changes would be extremely useful. For now, I listen for changes in my settings page, and take action there if needed (mainly if the user language changes), but broadcasting would be a much better pattern, more robust, and faster. If all components were listening for the broadcast and dealing with the settings changes that were relevant to them, it would be be fantastic.

**Comment by wbhob** [comment]
I don't think I can implement this, but if it works please make a PR!

**Comment by wbhob** [comment]
Max didn't like the idea of promises, so I restored it to its original state but implemented an observable instead.

**Comment by bgies** [comment]
Max doesn't like promises? BUT.. you put it back to the state when you return a promise EVERY TIME you get a value.... Not sure I understand that reasoning.... at the minimum you should implement something like ng2-translate's instant method.

**Comment by simonhaenisch** [comment]
After trying to figure out how to get a value from the settings and having issues with `load()` not returning default settings on first load (using sqlite/websql), I realized that this whole settings provider is quite a mess and found this issue.

I like the idea of @bgies and @bojanvidanovic about using an observable, is anyone working on a PR for that (PR #59 got closed)? Otherwise I'll create one.

**Comment by wbhob** [comment]
I've been meaning to reorganize #59. I'll probably get to it Monday or tuesday

**Comment by simonhaenisch** [comment]
I ended up writing this now because I needed it, so I'll just share it here for reference. I didn't really get why both `Observable` and `BehaviorSubject` would be needed, so I only used the latter.

I made one big change and that is using the `defaults` constructor parameter as the initial value for the `settings` member, and merging the settings from storage into them (overwriting) after they get retrieved. This works quite well in my case (rendering with default settings at initial app load and then re-rendering once the user's settings are loaded from storage). However in other cases it might not be desirable and you might want to await `storage.get()` before rendering (wouldn't be too hard, just add an `initialise()` method or something to read from storage and wait for the promise to resolve before rendering the app).

**`providers/settings.ts`:**

```ts
import { Injectable } from '@angular/core';
import { Storage } from '@ionic/storage';
import { BehaviorSubject } from 'rxjs';

/**
 * A simple settings/config class for storing key/value pairs with persistence.
 */
@Injectable()
export class Settings {
  private SETTINGS_STORAGE_KEY = 'user_settings';
  private settings: any = this.defaults;

  subject: BehaviorSubject<any> = new BehaviorSubject(this.settings);

  constructor(private storage: Storage, private defaults: any) {
    this.storage.get(this.SETTINGS_STORAGE_KEY).then(settings => {
      this.save(settings ? settings : null);
    });
  }

  private merge(settings: any, overwrite: boolean = true) {
    for (const k in settings) {
      if (overwrite || !(k in this.settings)) {
        this.settings[k] = settings[k];
      }
    }
  }

  getAll(): object {
    return this.settings;
  }

  getValue(key: string) {
    return this.settings[key];
  }

  setValue(key: string, value: any) {
    this.settings[key] = value;
    this.broadcast();
    return this.save();
  }

  save(settings: any = null) {
    if (settings) {
      this.merge(settings);
      this.broadcast();
    }

    return this.storage.set(this.SETTINGS_STORAGE_KEY, this.settings);
  }

  broadcast(settings = this.settings) {
    this.subject.next(settings);
  }
}
```

Also, because I did a bit of renaming, two or three lines in `pages/settings/settings.ts` might have to be adjusted:

* `settings.allSettings` becomes `settings.getAll` and doesn't return a promise anymore, thus replace [settings.ts#L84-L89](https://github.com/ionic-team/ionic-starter-super/blob/515fdae1ef5cf8737a0d3714f0aac03cb4bbf80d/src/pages/settings/settings.ts#L84-L89) with:
```ts
this.options = this.settings.getAll();
this.settingsReady = true;
```
* In [settings.ts#L63-L65](https://github.com/ionic-team/ionic-starter-super/blob/515fdae1ef5cf8737a0d3714f0aac03cb4bbf80d/src/pages/settings/settings.ts#L63-L65), use `settings.save()` instead of `settings.merge()`:
```ts
this.form.valueChanges.subscribe(() => {
  this.settings.save(this.form.value);
});
```

In a page you can subscribe to the settings subject like that:

```ts
this.settings.subject.subscribe(settings => {
  this.option = settings.option;
});
```

**Comment by cprior** [comment]
Many thanks for the updated code, @simonhaenisch !


I believe that the last code snippet should read

> In a page you can subscribe to the settings subject like that:
```

this.settings.subject.subscribe(settings => {
  this.option = settings.subject;
});
```

but I had also renamed a whole lot of stuff (and used a FormGroup).

**Comment by simonhaenisch** [comment]
What is `settings.subject` supposed to be? You could only access the `subject` key in `settings` if you stored something in that key, i. e. `setValue('subject', 'something')`. Or do you store a reference to the subject in the settings (that wouldn't make sense to me)?

I actually just worked on this again today, removed the `getAll()` and `getValue()` methods (only allowing access through subscriptions). And I also don't initialize the subject with the default settings anymore but an empty object instead. That makes it easier to check if the settings have been loaded from storage already. Might share some more code later.

**Comment by cprior** [comment]
Maybe I have changed the example to the worse! ;)

The line in https://github.com/ionic-team/ionic-starter-super/issues/58#issuecomment-313301886 containing
```
this.option = settings.option;
```
had an undefined settings.option for me, and I must have gotten mixed up in my correction.


Anyhow, I got your code working for me, many thanks for that!

**Comment by simonhaenisch** [comment]
Ah ok, that was just an example for how to read one option (called `option`) from the settings. Of course you would need to store something in the `option` key first 😉 

This is what I currently have in `providers/settings.ts`:

```ts
import { Injectable } from '@angular/core';
import { Storage } from '@ionic/storage';
import { BehaviorSubject } from 'rxjs';
import { STORAGE_KEYS } from '../app/config';

@Injectable()
export class Settings {
  constructor(public storage: Storage, public defaults: any) {
    // load settings from storage (async)
    this.storage.get(STORAGE_KEYS.SETTINGS).then(settings => {
      this.save(settings || null);
    });
  }

  // don't access settings directly, but subscribe to the subject instead
  // the subject will initially return/broadcast null (until loading finishes)
  // test if a key exists with `if ('key' in settings)`
  private settings = this.defaults;
  public subject: BehaviorSubject<any> = new BehaviorSubject(null);

  // merge object into settings (defaults to overwrite existing)
  merge(settings: any, overwrite = true) {
    for (const key in settings) {
      if (overwrite || !(key in this.settings)) {
        this.settings[key] = settings[key];
      }
    }
  }

  // set value, then save to storage (which also broadcasts it)
  // (await this to make sure it got saved)
  setValue(key: string, value: any) {
    this.settings[key] = value;
    return this.save();
  }

  // optionally merge the passed object into settings, then broadcast and save
  // (await this to make sure it got saved)
  save(settings: any = null) {
    if (settings) {
      this.merge(settings);
    }

    this.broadcast();

    return this.storage.set(STORAGE_KEYS.SETTINGS, this.settings);
  }

  // broadcast settings to everyone who subscribes to the subject
  broadcast(settings = this.settings) {
    this.subject.next(settings);
  }
}
```

And then in a page:

```ts
export class Page {
  constructor(public settings: Settings) {
    this.settings.subject.subscribe(settings => {
      if (settings) {
        console.log(settings));
      }
    }
  }
}
```
